### ISSUE TITLE
Using soft prediction

### ISSUE THREAD

**Issue by rose-jinyang** [PeikeLi/Self-Correction-Human-Parsing#62]
Hello
How are you?
Thanks for contributing to this project.
I am training a new model with your project.
But I think that there is an issue in the below part of "train.py" script.

![image](https://user-images.githubusercontent.com/47862419/132393806-9897bf89-078c-4dea-a72c-d622a9b0e788.png)

What about revising this part as the following?

![image](https://user-images.githubusercontent.com/47862419/132393929-97c4ee40-06e3-4980-9932-46cb99e0caf3.png)

**Comment by KudoKhang** [comment]
<img width="590" alt="image" src="https://user-images.githubusercontent.com/24488797/171212677-2e40db79-ef52-48e4-a906-e3c93e16f13f.png">

That's right, it has a problem. And here's my code

**Comment by jackylu0124** [comment]
> <img alt="image" width="590" src="https://user-images.githubusercontent.com/24488797/171212677-2e40db79-ef52-48e4-a906-e3c93e16f13f.png">
> 
> That's right, it has a problem. And here's my code

Could you please explain the purpose/reasoning behind the for-loop in the code you showed in the screenshot? I think rose's solution makes more sense because in the code you provided, it would only work when the batch size is set to 2 and won't work for other batch size numbers since the two tensors passed into `moving_average()` need to have the same dimensions. I think the correct solution should be something like the one below, which is essentially rose's version. Let me know what you think, thanks!

```
soft_parsing = soft_preds[0][-1]
soft_edges = soft_preds[1][-1]
soft_preds = soft_parsing
```

**Comment by duACGN** [comment]
> > <img alt="image" width="590" src="https://user-images.githubusercontent.com/24488797/171212677-2e40db79-ef52-48e4-a906-e3c93e16f13f.png">
> > That's right, it has a problem. And here's my code
> 
> Could you please explain the purpose/reasoning behind the for-loop in the code you showed in the screenshot? I think rose's solution makes more sense because in the code you provided, it would only work when the batch size is set to 2 and won't work for other batch size numbers since the two tensors passed into `moving_average()` need to have the same dimensions. I think the correct solution should be something like the one below, which is essentially rose's version. Let me know what you think, thanks!
> 
> ```
> soft_parsing = soft_preds[0][-1]
> soft_edges = soft_preds[1][-1]
> soft_preds = soft_parsing
> ```

Thank you for the method you provided, but I trained LIP data in this way. The mIoU does not reach 58.62 of the author's paper. In fact, I trained only 57.98 mIoU. Do you have any better method?

**Comment by MrAriten** [comment]
> > > <img alt="image" width="590" src="https://user-images.githubusercontent.com/24488797/171212677-2e40db79-ef52-48e4-a906-e3c93e16f13f.png">
> > > That's right, it has a problem. And here's my code
> > 
> > 
> > Could you please explain the purpose/reasoning behind the for-loop in the code you showed in the screenshot? I think rose's solution makes more sense because in the code you provided, it would only work when the batch size is set to 2 and won't work for other batch size numbers since the two tensors passed into `moving_average()` need to have the same dimensions. I think the correct solution should be something like the one below, which is essentially rose's version. Let me know what you think, thanks!
> > ```
> > soft_parsing = soft_preds[0][-1]
> > soft_edges = soft_preds[1][-1]
> > soft_preds = soft_parsing
> > ```
> 
> Thank you for the method you provided, but I trained LIP data in this way. The mIoU does not reach 58.62 of the author's paper. In fact, I trained only 57.98 mIoU. Do you have any better method?

Are your hyperparameter settings such as batchsize and learning rate the same as those of the author?
